pkg/resourcemonitor/podresourcesscanner.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcemonitor
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	podresourcesapi "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"sigs.k8s.io/node-feature-discovery/pkg/apihelper"
+)
+
+type PodResourcesScanner struct {
+	namespace         string
+	podResourceClient podresourcesapi.PodResourcesListerClient
+	apihelper         apihelper.APIHelpers
+}
+
+func NewPodResourcesScanner(namespace string, podResourceClient podresourcesapi.PodResourcesListerClient, kubeApihelper apihelper.APIHelpers) (ResourcesScanner, error) {
+	resourcemonitorInstance := &PodResourcesScanner{
+		namespace:         namespace,
+		podResourceClient: podResourceClient,
+		apihelper:         kubeApihelper,
+	}
+	if resourcemonitorInstance.namespace != "" {
+		klog.Infof("watching namespace %q", resourcemonitorInstance.namespace)
+	} else {
+		klog.Infof("watching all namespaces")
+	}
+
+	return resourcemonitorInstance, nil
+}
+
+// isWatchable tells if the the given namespace should be watched.
+func (resMon *PodResourcesScanner) isWatchable(podNamespace string, podName string, hasDevice bool) (bool, bool, error) {
+	cli, err := resMon.apihelper.GetClient()
+	if err != nil {
+		return false, false, err
+	}
+	pod, err := resMon.apihelper.GetPod(cli, podNamespace, podName)
+	if err != nil {
+		return false, false, err
+	}
+
+	klog.Infof("podresource: %s", podName)
+	isIntegralGuaranteed := podGuaranteedCPUs(pod)
+
+	if resMon.namespace == "" && (isIntegralGuaranteed || hasDevice) {
+		return true, isIntegralGuaranteed, nil
+	}
+	// TODO:  add an explicit check for guaranteed pods and pods with devices
+	return resMon.namespace == podNamespace && (isIntegralGuaranteed || hasDevice), isIntegralGuaranteed, nil
+}
```

Comment:
This whole function is quite murky to me. As I suggested earlier, I think we should always watch all namespaces. Also, shouldn't we account for burstable containers, too?

Reply:
Please refer to the comment [here](https://github.com/kubernetes-sigs/node-feature-discovery/pull/525#discussion_r700487635) that explains why we are supporting ability to provide configurable namespaces. Let me explain what we are trying to do in this function:
The goal is to determine if a particular pod is watchable and should be decremented from the capacity per NUMA. There are essentially two aspects to this:
1. if the watch-namespace= "" (which corresponds to all the namespaces) we consider the pod if the pod is either allocated exclusive CPUs or is allocated devices (because if there is pod that is not requesting CPUs but getting devices we want to track them).
2.  if the watch-namespace is specified to be a specific namespace (mostly for testing purpose) we check if a pod belongs to that namespace and then essentially again check for the same thing as we checked in the previous case:  consider the pod if it is either allocated exclusive CPUs or is allocated devices.
Hope this helps. Let me know if you need further clarification.

Reply:
> 2\. if the watch-namespace is specified to be a specific namespace (mostly for testing purpose)

OK, you could probably add a mention that this is meant for testing/debugging purposes e.g. in the flag description. WDYT?

> If a pod is allocated devices irrespective of its QoS we consider that pod because devices are always exclusively allocated to a pod. However, in case of CPUs because burstable pods obtain CPUs from shared pool there is no way to determine NUMA nodes corresponding so we don't consider such pods.

Ah, yes as we're only tracking cpu ids